---
title: "Cell cell communication in COVID data"
author: "Yue Cao"
date: "2025-3-26"
output: md_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)



.libPaths("/dora/nobackup/yuec/R")
library(SingleCellExperiment)
library(ggplot2)
library(ggpubr)
library(ggthemes)
library(CellChat)
library(reshape2)
library(cluster)
library(RColorBrewer)
library(gridExtra)
library(scales)
library(stringr)
library(pheatmap)
library(scattermore)

theme_set(theme_bw() +
               theme(panel.grid.major = element_blank(), 
                     panel.grid.minor = element_blank(),
                     panel.background = element_rect(colour = "black", size = 1),
                     axis.text.x = element_text(color = "black", angle = 45, hjust = 1),
                     axis.text.y = element_text(color = "black")))

```


# Introduction 


In this analysis, we examine a COVID-19 single-cell RNA-sequencing data obtained from the study " COVID-19 severity correlates with airway epitheliumâ€“immune cell interactions identified by single-cell analysis" by Chua et al., published in Nature Biotechnology. The paper sequenced nasopharyngeal and bronchial samples to examine molecular mechanisms behind COVID-19 disease severity.   


In this analysis, we aim to perform some basic data exploration follow by cell cell interaction analysis. Cell cell interaction infers the communication between individual cells. Given a single-cell RNA-sequencing data, it predicts the interaction between cell types and between ligand receptor pairs based on the gene expression of individual cells.  


# Basic data exploration 

First, we subset the data to moderate and critical patients and explore the dataset. 
 
 
```{r include=FALSE}
load_data <- function(){
  
  chua  <- readRDS("/dski/nobackup/biostat/datasets/singlecell/COVID19/sce_chua_20200721.rds")
  
  return(chua)
}
```
 
 
```{r}

chua <- load_data() 

chua <- chua[, chua$severity %in% c("moderate" , "critical")]

meta_chua <- colData(chua)
meta_chua  <- meta_chua [!duplicated(meta_chua$sample), ]
rownames(meta_chua) <- meta_chua$sample

meta_chua_unique <- meta_chua[!duplicated(meta_chua$sample), ]

print("dimension of the data")
dim( chua) 

print("number of samples in each severity class")
table(as.character( meta_chua_unique$severity) ) 


```


Visualising the cell type composition of the samples. There are 13 cell types in total, B, Basal, Basophil/Mast , Ciliated , Dendritic, Fibroblast, Goblet, intermediate, Macrophage, Monocyte, Neutrophil, T , unassigned. Neurophil, goblet, T cells are some of the major cell types.   


```{r fig.height=6, fig.width=8}

df <- data.frame( table(chua$sample, chua$pred_chua_scClassify) )
 
colnames(df) <- c("sample" , "celltype" , "count" )

df  <- suppressMessages(  df  %>%
  dplyr::group_by( sample, celltype) %>%
  dplyr::summarise(Total = sum(count)) %>%
  dplyr::mutate(Proportion = Total / sum(Total)) ) 

df$severity <- chua[ , match(df$sample, chua$sample )]$severity

 
ggplot( df, aes(x = sample, y = Proportion, fill = celltype)) + geom_col()  + facet_wrap(~severity,scale = "free") + scale_fill_tableau("Tableau 20")


```

 



# Exploring cell cell communication  

Now, we perform cell cell communication algorithm and explore the results. We first ran cell cell communication analysis using a popular method called CellChat. CellChat infers the cell cell communication within each individual. We then aggregates the outputs within moderate and within critical samples to compare the overall profile of the moderate and the critical samples.  


```{r eval=FALSE, echo=FALSE}


run_CCI <- function(){
  future::plan("multisession", workers = 1)
  
  # normalised gene expression matrix (gene by cell)
  exprsMat <- logcounts(chua)
  # meta info for each cell
  meta <-  colData(chua)
  # The cell type info used as cell grouping
  celltype_info <- meta$pred_chua_scClassify
  # The sample (patient) info used as cell grouping
  patient_info <- meta$sample
  
  
  
  sce <- SingleCellExperiment(assay = list(logcounts = exprsMat),
                              colData = meta)
  
  sce$celltype <- celltype_info
  sce$patient <- patient_info
  tab <- table(sce$patient)
  sample_name <- names(tab)[tab >= 100]
  
  
  s <- 2
  
  result_list <- mclapply( 1:length(sample_name), function(s) {
  
    print(sample_name[s])
  
    sce_patient <- sce[, sce$patient == sample_name[s]]
    sce_patient <- sce_patient[, !sce_patient$celltype %in%
                                 c("intermediate", "unassigned")]
  
  
    tab <- table(sce_patient$celltype)
    print(tab)
    sce_patient <- sce_patient[, sce_patient$celltype %in% names(which(tab > 5))]
    print(table(sce_patient$celltype))
  
    if (length(table(sce_patient$celltype)) > 3) {
  
  
      cellchat <- createCellChat(object = as.matrix(logcounts(sce_patient)),
                                 meta = data.frame(colData(sce_patient)),
                                 group.by = "celltype")
  
      cellchat <- setIdent(cellchat, ident.use = "celltype") # set "labels" as default cell identity
  
  
      levels(cellchat@idents) # show factor levels of the cell labels
      groupSize <- as.numeric(table(cellchat@idents)) # number of cells in each cell group
  
      CellChatDB <- CellChatDB.human # use CellChatDB.human if running on human data
      showDatabaseCategory(CellChatDB)
      # Show the structure of the database
      dplyr::glimpse(CellChatDB$interaction)
  
      CellChatDB.use <- CellChatDB # use Secreted Signaling for cell-cell communication analysis
      cellchat@DB <- CellChatDB.use # set the used database in the object
  
      cellchat <- subsetData(cellchat) # subset the expression data of signaling genes for saving computation cost
      # do parallel
      cellchat <- identifyOverExpressedGenes(cellchat)
      cellchat <- identifyOverExpressedInteractions(cellchat)
      cellchat <- projectData(cellchat, PPI.human)
      cellchat <- computeCommunProb(cellchat)
      cellchat <- computeCommunProbPathway(cellchat)
      cellchat <- aggregateNet(cellchat)
  
  
    }
  
  }, mc.cores = 35)
  
  names(result_list) <- sample_name
  
  saveRDS(result_list , "chua_cellchat_list.rds")

  
}

```
 


```{r eval=FALSE}
cellchat_res_list <- run_CCI()
```
 
 
```{r echo=FALSE}

cellchat_res_list   <-  readRDS("/dski/nobackup/yuec/llm_scfeatures_report/CCI/chua_cellchat_list.rds")

cellchat_res_list <- cellchat_res_list [!sapply(cellchat_res_list , is.null)]

```
 
 
```{r echo=FALSE}


summarise_CCI_sample <- function( cellchat_res_list ){
  
  
   
       
      rankNet_byCellType <- function(object, slot.name = "netP", 
                                     x.rotation = 90, title = NULL, color.use = NULL, 
                                     bar.w = 0.75, font.size = 8) 
      {
          object1 <- methods::slot(object, slot.name)
          prob1 = object1$prob
          df <-  melt(apply(prob1, 3, function(x) {
              df <- suppressMessages( melt(x) )
              colnames(df) <- c("Ligand", "Receptor", "value")
              df
          })) 
          df <- df[, c("Ligand", "Receptor", "L1", "value")]
          colnames(df)[3] <- "Pathway"
          return(df)
          
          
      }
      
      dcast2 <- function (x, ...) {
        data <-  reshape2::dcast(x, ...)
        rownames(data) <- data[, 1]
        data <- data[, -1]
        return(data)
      }
      
      
      rankNet_byCellType_list <- suppressMessages( lapply(cellchat_res_list, rankNet_byCellType) )
      
       
      rankNet_byCellType_list <-  suppressMessages( melt(rankNet_byCellType_list) )
      rankNet_byCellType_list$Ligand_group <- unlist(lapply(strsplit(as.character(rankNet_byCellType_list$Ligand), "_"), "[[", 1))
      rankNet_byCellType_list$Receptor_group <- unlist(lapply(strsplit(as.character(rankNet_byCellType_list$Receptor), "_"), "[[", 1))
      
      
       
      rankNet_byGroup_agg <- aggregate(rankNet_byCellType_list$value, 
                                       list(rankNet_byCellType_list$Ligand_group,
                                            rankNet_byCellType_list$Receptor_group,
                                            rankNet_byCellType_list$L1,
                                            rankNet_byCellType_list$Pathway),
                                       sum)
      
      
      colnames(rankNet_byGroup_agg) <- c("Ligand_group", 
                                         "Receptor_group",
                                         "sample",
                                         "Pathway",
                                         "value")
      features <- paste(rankNet_byGroup_agg$Ligand_group,
                        rankNet_byGroup_agg$Receptor_group,
                        rankNet_byGroup_agg$Pathway, sep = "_")
      
      rankNet_byGroup_agg$features <- features
      
      
        
      aff_mat_bySample <- lapply(split(rankNet_byGroup_agg, rankNet_byGroup_agg$sample),
                                 function(x) dcast2(x, Ligand_group~Receptor_group,
                                                    fun.aggregate = sum, value.var = "value"))
      all_cellTypes <- names(table(rankNet_byGroup_agg$Ligand_group))
      
      
      aff_mat_bySample <- lapply(aff_mat_bySample, function(x) {
          mat <- matrix(0, ncol = length(all_cellTypes), nrow = length(all_cellTypes))
          colnames(mat) <- rownames(mat) <- all_cellTypes
          mat[rownames(x), colnames(x)] <- as.matrix(x)
          mat
      })
      
       
      aff_mat_bySample <- lapply(aff_mat_bySample, function(x) {
          (x - min(x))/(max(x) - min(x))
      })


return( aff_mat_bySample)
}


```



```{r warning=FALSE, message=FALSE}

 
aff_mat_bySample <- summarise_CCI_sample(cellchat_res_list )

 
critical_patients <- rownames(meta_chua)[meta_chua$severity == "critical"]

aff_mat_critical <- Reduce("+", aff_mat_bySample[names(aff_mat_bySample) %in% critical_patients])/length(critical_patients)

 
moderate_patients <- rownames(meta_chua)[meta_chua$severity == "moderate"]

aff_mat_moderate <- Reduce("+", aff_mat_bySample[names(aff_mat_bySample) %in% moderate_patients])/length(moderate_patients)
 

```



## Overall cell cell communication pattern between moderate and critical patients 


After we obtain the output from the algorithm, we first examine the cell cell communication patterns between pairwise cell types and compare the patterns between moderate and critical samples.     

The plot below visualises the difference between moderate and critical samples, where red indicates a stronger signal in critical samples, blue indicates greater signal in moderate samples.    

For example, in the mild patients, we see there is a stronger interaction between Goblet (sender cell type) with T (receiver cell type). In the critical patients, there is a stronger interaction between Monocyte (sender cell type) with neutrophil (receiver cell type).    
 
 
```{r fig.height=6, fig.width=8}

mat <- aff_mat_critical - aff_mat_moderate
 


melted_data <- suppressMessages( melt(as.matrix(mat)) )
colnames(melted_data) <- c("Row", "Column", "Value")


ggplot(melted_data, aes(x = Column, y = Row, fill = Value)) +
  geom_tile() +
  scale_fill_gradient2(
    low = "#2166ac",
    mid = "white",
    high = "#b2182b",
    midpoint = 0,
    limits = c(min(melted_data$Value), max(melted_data$Value))
  ) +
  labs(fill = "Cell cell communication") +    geom_text(aes(label = round(Value,2)), color = "black", size = 3 ) + ylab("Source (sender)") + xlab("Target (receiver)") + ggtitle("Critical vs moderate")




```
 
## Ligand receptor in the mild patients


Here we examine the top ligand receptor interactions in the mild patients. From the previous result, we noticed there is a greater interaction in the mild patients with goblet being the sender cell type. Therefore, we selectively focus on goblet being the sender, and macrophage, monocyte, neutrophil and T cells being the receiver and examine the top ligand receptor within these cell type pairs. We show the top 20 ligand receptor interactions, ranked by the interaction probability. 


```{r include=FALSE}

get_ligand_receptor <- function( cellchat_res_list ){
  
      thissample <-  names(cellchat_res_list)[1]
    
    ligand_receptor <- NULL
    
    for ( thissample in names(cellchat_res_list) ){
      
      thisresult <- suppressMessages( netVisual_bubble(cellchat_res_list[[thissample]],  return.data = T))
      thisresult <- thisresult$communication
      thisresult$sample <- thissample
     
      ligand_receptor  <- rbind(ligand_receptor ,  thisresult )
    }
    
    return(ligand_receptor)

}

 

```

```{r}

ligand_receptor <- get_ligand_receptor(cellchat_res_list )

ligand_receptor_moderate <-  ligand_receptor[ ligand_receptor$source %in% c("Goblet" ) & 
                ligand_receptor$target %in%  c(  "Macrophage",  "Monocyte", "Neutrophil",  "T"),  ]

 
ligand_receptor_moderate <-  ligand_receptor_moderate[ ligand_receptor_moderate$sample %in% moderate_patients,  ]
 
ligand_receptor_moderate <-  suppressMessages(  ligand_receptor_moderate %>%
        dplyr::group_by( source.target, interaction_name_2) %>%
        dplyr::summarise(prob = mean(prob)) ) 


data.frame( ligand_receptor_moderate [order(ligand_receptor_moderate$prob , decreasing = T ) , ][ 1:20, ] )


```


## ligand receptor in the critical patients 

Here we examine the top ligand receptor interactions in the critical patients. From the previous result, we noticed there is a greater interaction in the critical patients with monocyte being the sender cell type and neutrophil being the receiver. Therefore, we examine the top ligand receptor within these cell type pairs. We show the top 20 ligand receptor interactions, ranked by the interaction probability. 



```{r}

ligand_receptor_critical <-  ligand_receptor[ ligand_receptor$source %in% c("Monocyte" ) & 
                ligand_receptor$target %in%  c(  "Neutrophil"),  ]


ligand_receptor_critical <-  ligand_receptor_critical[ ligand_receptor_critical$sample %in% critical_patients,  ]
 
ligand_receptor_critical <- suppressMessages(  ligand_receptor_critical %>%
        dplyr::group_by( source.target, interaction_name_2) %>%
        dplyr::summarise(prob = mean(prob)) )


data.frame( ligand_receptor_critical[order(ligand_receptor_critical$prob , decreasing = T ) , ][ 1:20, ] ) 


```

